kernel/kernel-api/src/main/java/io/delta/kernel/internal/compaction/LogCompactionWriter.java

```diff
@@ -0,0 +1,162 @@
+import io.delta.kernel.internal.util.FileNames.DeltaLogFileType;
+import io.delta.kernel.utils.CloseableIterator;
+import io.delta.kernel.utils.FileStatus;
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Utility for writing out log compactions. */
+public class LogCompactionWriter {
+
+  private static final Logger logger = LoggerFactory.getLogger(LogCompactionWriter.class);
+
+  private final Path logPath;
+  private final long startVersion;
+  private final long endVersion;
+  private final long minFileRetentionTimestampMillis;
+
+  public LogCompactionWriter(
+      Path logPath, long startVersion, long endVersion, long minFileRetentionTimestampMillis) {
+    this.logPath = requireNonNull(logPath);
+    this.startVersion = startVersion;
+    this.endVersion = endVersion;
+    this.minFileRetentionTimestampMillis = minFileRetentionTimestampMillis;
+  }
+
+  public void writeLogCompactionFile(Engine engine) throws IOException {
+    String fileName = String.format("%020d.%020d.compacted.json", startVersion, endVersion);
+    Path compactedPath = new Path(logPath, fileName);
+
+    logger.info(
+        "Writing compacted log file for versions {} to {} to path: {}",
+        startVersion,
+        endVersion,
+        compactedPath);
+
+    final long startTimeMillis = System.currentTimeMillis();
+    final List<FileStatus> deltas =
+        DeltaLogActionUtils.listDeltaLogFilesAsIter(
+                engine,
+                new HashSet<>(Arrays.asList(DeltaLogFileType.COMMIT)),
+                logPath,
+                startVersion,
+                Optional.of(endVersion),
+                true /* mustBeRecreatable */)
+            .toInMemoryList();
+
+    logger.info(
+        "{}: Took {}ms to list commit files for log compaction",
+        logPath,
+        System.currentTimeMillis() - startTimeMillis);
+
+    if (deltas.isEmpty()) {
+      logger.warn(
+          "Asked to do a log compaction between {} and {}, " + "but there are no files to compact",
+          startVersion,
+          endVersion);
+      return;
+    }
+
+    final long lastCommitTimestamp = getLast(deltas).getModificationTime();
+
+    LogSegment segment =
+        new LogSegment(logPath, endVersion, deltas, emptyList(), lastCommitTimestamp);
+    CreateCheckpointIterator checkpointIterator =
+        new CreateCheckpointIterator(engine, segment, minFileRetentionTimestampMillis);
+    wrapEngineExceptionThrowsIO(
+        () -> {
+          try (CloseableIterator<Row> rows = new FilteredBatchToRowIter(checkpointIterator)) {
+            engine.getJsonHandler().writeJsonFileAtomically(compactedPath.toString(), rows, false);
+          }
+          logger.info("Successfully wrote compacted log file `{}`", compactedPath);
+          return null;
+        },
+        "Wrote compacted log file `%s`",
```

Comment:
I think the verbiage of this should be what we were doing as opposed to what the successful outcome should have been if we completed -- this is for an error, after all.

`Writing log compaction file ...`